### PR TITLE
protocol-classification: Improving classification according to the RFC

### PR DESCRIPTION
**Pull request by guyarb**
### What does this PR do?

Improving HTTP2 classification by checking if the server returns a settings frame.

### Motivation

Improving the classification of HTTP2 to be more accurate

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 1]
```diff
@@ -35,6 +36,21 @@ static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
     return match;
 }
 
+// According to the https://www.rfc-editor.org/rfc/rfc7540#section-3.5
+// a HTTP2 server must reply with a settings frame to the preface of HTTP2.
```
Comment (DanielLavie) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:40]:
```suggestion
// an HTTP2 server must reply with a settings frame to the preface of HTTP2.
```

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 2]
```diff
@@ -75,7 +91,7 @@ static __always_inline void classify_protocol(protocol_t *protocol, const char *
 
     if (is_http(buf, size)) {
         *protocol = PROTOCOL_HTTP;
-    } else if (is_http2(buf, size)) {
+    } else if (is_http2(buf, size) || is_http2_server_settings(buf, size)) {
```
Comment (DanielLavie) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:94]: Consider moving the call to `is_http2_server_settings` to keep this function cleaner

**Review by DanielLavie — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h (hunk 2 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:94]: changed

**Review by DanielLavie — APPROVED** [review]

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 3]
```diff
@@ -8,33 +8,55 @@
...
+    if (!read_http2_frame_header(buf, buf_size, &frame_header)) {
+        return false;
+    }
+
+    return frame_header.type == kSettingsFrame && frame_header.stream_id == 0 && frame_header.length % HTTP2_SETTINGS_SIZE == 0;
+}
+
+// The method checks if the given buffer starts with the HTTP2 marker as defined in https://datatracker.ietf.org/doc/html/rfc7540.
+// We check that the given buffer is not empty and its size is at least 24 bytes.
+static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
```
Comment (hmahmood) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:56]: Is this getting called from somewhere?

Comment (guyarb) [reply pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:56]: yes, https://github.com/DataDog/datadog-agent/blob/guy.arbitman/USMO-146-improve-http2-classification/pkg/network/ebpf/c/protocols/protocol-classification-helpers.h#L100

Comment (guyarb) [reply pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:56]: namely `classify_protocol` function

**Review by hmahmood — APPROVED** [review]

**Pull request merged**